File: code/06-search-trees/python/binary_search_tree.py
"""BinarySearchTree: distinct int key, parent pointer, SEARCH/MINIMUM/MAXIMUM/
SUCCESSOR/PREDECESSOR/INSERT/DELETE(Transplant 기반) 전부 O(h).
lecture-notes/code/lecture06/java/BinarySearchTree.java와 같은 정책·예제 입력을 쓴다."""


# snippet:binary-search-tree:start
class _Node:
    def __init__(self, key):
        self.key = key
        self.left = self.right = self.parent = None


class BinarySearchTree:
    def __init__(self):
        self.root = None
        self._size = 0

    def __len__(self):
        return self._size

    def _find(self, key):
        x = self.root
        while x is not None and x.key != key:
            x = x.left if key < x.key else x.right
        return x

    def contains(self, key):
        return self._find(key) is not None

    def insert(self, key):
        y, x = None, self.root
        while x is not None:
            y = x
            if key == x.key:
                return False
            x = x.left if key < x.key else x.right
        z = _Node(key)
        z.parent = y
        if y is None:
            self.root = z
        elif key < y.key:
            y.left = z
        else:
            y.right = z
        self._size += 1
        return True

    @staticmethod
    def _minimum(x):
        if x is None:
            return None
        while x.left is not None:
            x = x.left
        return x

    @staticmethod
    def _maximum(x):
        if x is None:
            return None
        while x.right is not None:
            x = x.right
        return x

    def minimum(self):
        x = self._minimum(self.root)
        return None if x is None else x.key

    def maximum(self):
        x = self._maximum(self.root)
        return None if x is None else x.key

    @staticmethod
    def _successor(x):
        if x.right is not None:
            return BinarySearchTree._minimum(x.right)
        y = x.parent
        while y is not None and x is y.right:
            x, y = y, y.parent
        return y

    @staticmethod
    def _predecessor(x):
        if x.left is not None:
            return BinarySearchTree._maximum(x.left)
        y = x.parent
        while y is not None and x is y.left:
            x, y = y, y.parent
        return y

    def successor(self, key):
        x = self._find(key)
        if x is None:
            return None
        y = self._successor(x)
        return None if y is None else y.key

    def predecessor(self, key):
        x = self._find(key)
        if x is None:
            return None
        y = self._predecessor(x)
        return None if y is None else y.key

    def _transplant(self, u, v):
        if u.parent is None:
            self.root = v
        elif u is u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        if v is not None:
            v.parent = u.parent

    def delete(self, key):
        z = self._find(key)
        if z is None:
            return False
        if z.left is None:
            self._transplant(z, z.right)
        elif z.right is None:
            self._transplant(z, z.left)
        else:
            y = self._minimum(z.right)
            if y.parent is not z:
                self._transplant(y, y.right)
                y.right = z.right
                y.right.parent = y
            self._transplant(z, y)
            y.left = z.left
            y.left.parent = y
        self._size -= 1
        return True

    def inorder(self):
        out = []

        def walk(x):
            if x is not None:
                walk(x.left)
                out.append(x.key)
                walk(x.right)

        walk(self.root)
        return out
# snippet:binary-search-tree:end

if __name__ == "__main__":
    t = BinarySearchTree()
    def fmt(v):
        return "none" if v is None else str(v)

    for k in [15, 6, 3, 2, 4, 7, 13, 9, 14, 18, 17, 20]:
        t.insert(k)
    print("inorder:", ",".join(str(k) for k in t.inorder()))
    t.insert(12)
    print("insert12 inorder:", ",".join(str(k) for k in t.inorder()))
    print("minimum:", fmt(t.minimum()))
    print("maximum:", fmt(t.maximum()))
    print("successor15:", fmt(t.successor(15)))
    print("successor6:", fmt(t.successor(6)))
    print("successor4:", fmt(t.successor(4)))
    print("successor20:", fmt(t.successor(20)))
    print("predecessor15:", fmt(t.predecessor(15)))
    t.delete(6)
    print("delete6 inorder:", ",".join(str(k) for k in t.inorder()))
    t.delete(15)
    print("delete15 inorder:", ",".join(str(k) for k in t.inorder()))

File: code/06-search-trees/java/BinarySearchTree.java
import java.util.ArrayList;
import java.util.List;
import java.util.OptionalInt;

/** lecture-notes/code/lecture06/java/BinarySearchTree.java와 동일(verbatim).
 * Distinct int keys, parent pointers, all operations O(h). */
public final class BinarySearchTree {
    // snippet:binary-search-tree:start
    private static final class Node {
        int key; Node left, right, parent;
        Node(int key) { this.key = key; }
    }
    private Node root;
    private int size;
    public int size() { return size; }
    private Node find(int key) {
        Node x=root;
        while(x!=null && x.key!=key) x=key<x.key?x.left:x.right;
        return x;
    }
    public boolean contains(int key) { return find(key)!=null; }
    public boolean insert(int key) {
        Node y=null,x=root;
        while(x!=null){y=x;if(key==x.key)return false;x=key<x.key?x.left:x.right;}
        Node z=new Node(key);z.parent=y;
        if(y==null)root=z;else if(key<y.key)y.left=z;else y.right=z;
        size++;return true;
    }
    private static Node minimum(Node x){if(x==null)return null;while(x.left!=null)x=x.left;return x;}
    private static Node maximum(Node x){if(x==null)return null;while(x.right!=null)x=x.right;return x;}
    public OptionalInt minimum(){Node x=minimum(root);return x==null?OptionalInt.empty():OptionalInt.of(x.key);}
    public OptionalInt maximum(){Node x=maximum(root);return x==null?OptionalInt.empty():OptionalInt.of(x.key);}
    private static Node successor(Node x){
        if(x.right!=null)return minimum(x.right);
        Node y=x.parent;while(y!=null&&x==y.right){x=y;y=y.parent;}return y;
    }
    private static Node predecessor(Node x){
        if(x.left!=null)return maximum(x.left);
        Node y=x.parent;while(y!=null&&x==y.left){x=y;y=y.parent;}return y;
    }
    public OptionalInt successor(int key){Node x=find(key);if(x==null)return OptionalInt.empty();Node y=successor(x);return y==null?OptionalInt.empty():OptionalInt.of(y.key);}
    public OptionalInt predecessor(int key){Node x=find(key);if(x==null)return OptionalInt.empty();Node y=predecessor(x);return y==null?OptionalInt.empty():OptionalInt.of(y.key);}
    private void transplant(Node u,Node v){
        if(u.parent==null)root=v;else if(u==u.parent.left)u.parent.left=v;else u.parent.right=v;
        if(v!=null)v.parent=u.parent;
    }
    public boolean delete(int key){
        Node z=find(key);if(z==null)return false;
        if(z.left==null)transplant(z,z.right);
        else if(z.right==null)transplant(z,z.left);
        else{Node y=minimum(z.right);if(y.parent!=z){transplant(y,y.right);y.right=z.right;y.right.parent=y;}transplant(z,y);y.left=z.left;y.left.parent=y;}
        size--;return true;
    }
    public List<Integer> inorder(){List<Integer>o=new ArrayList<>();inorder(root,o);return List.copyOf(o);}
    private static void inorder(Node x,List<Integer>o){if(x!=null){inorder(x.left,o);o.add(x.key);inorder(x.right,o);}}
    // snippet:binary-search-tree:end

    private static String fmt(OptionalInt v) { return v.isPresent() ? String.valueOf(v.getAsInt()) : "none"; }
    private static String join(List<Integer> a) {
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < a.size(); i++) { if (i > 0) sb.append(','); sb.append(a.get(i)); }
        return sb.toString();
    }

    public static void main(String[] args) {
        BinarySearchTree t = new BinarySearchTree();
        for (int k : new int[] {15, 6, 3, 2, 4, 7, 13, 9, 14, 18, 17, 20}) t.insert(k);
        System.out.println("inorder: " + join(t.inorder()));
        t.insert(12);
        System.out.println("insert12 inorder: " + join(t.inorder()));
        System.out.println("minimum: " + fmt(t.minimum()));
        System.out.println("maximum: " + fmt(t.maximum()));
        System.out.println("successor15: " + fmt(t.successor(15)));
        System.out.println("successor6: " + fmt(t.successor(6)));
        System.out.println("successor4: " + fmt(t.successor(4)));
        System.out.println("successor20: " + fmt(t.successor(20)));
        System.out.println("predecessor15: " + fmt(t.predecessor(15)));
        t.delete(6);
        System.out.println("delete6 inorder: " + join(t.inorder()));
        t.delete(15);
        System.out.println("delete15 inorder: " + join(t.inorder()));
    }
}
